core/src/test/scala/unit/kafka/raft/KafkaNetworkChannelTest.scala

```diff
@@ -0,0 +1,252 @@
+
+    channel.send(request)
+    val responses = channel.receive(1000).asScala
+    assertEquals(1, responses.size)
+
+    val response = responses.head.asInstanceOf[RaftResponse.Inbound]
+    assertEquals(destinationId, response.sourceId)
+    assertEquals(requestId, response.requestId)
+    assertEquals(apiKey, ApiKeys.forId(response.data.apiKey))
+    assertEquals(error, extractError(response.data))
+  }
+
+  private def buildTestRequest(key: ApiKeys): ApiMessage = {
+    key match {
+      case ApiKeys.BEGIN_QUORUM_EPOCH =>
+        new BeginQuorumEpochRequestData()
+          .setClusterId(clusterId)
+          .setLeaderEpoch(5)
+          .setLeaderId(1)
+
+      case ApiKeys.END_QUORUM_EPOCH =>
+        new EndQuorumEpochRequestData()
+          .setClusterId(clusterId)
+          .setLeaderEpoch(5)
+          .setReplicaId(1)
+          .setLeaderId(1)
+
+      case ApiKeys.VOTE =>
+        new VoteRequestData()
+          .setClusterId(clusterId)
+          .setCandidateEpoch(5)
+          .setCandidateId(1)
+          .setLastEpoch(4)
+          .setLastEpochEndOffset(329)
+
+      case ApiKeys.FETCH_QUORUM_RECORDS =>
+        new FetchQuorumRecordsRequestData()
+          .setClusterId(clusterId)
+          .setLeaderEpoch(5)
+          .setReplicaId(1)
+          .setFetchOffset(333)
+          .setLastFetchedEpoch(5)
+
+      case ApiKeys.FIND_QUORUM =>
+        new FindQuorumRequestData()
+          .setHost("localhost")
+          .setPort(9092)
+          .setReplicaId(1)
+          .setBootTimestamp(time.milliseconds())
+
+      case _ =>
+        throw new AssertionError(s"Unexpected api $key")
+    }
+  }
+
+  private def buildTestErrorResponse(key: ApiKeys, error: Errors): ApiMessage = {
+    key match {
+      case ApiKeys.BEGIN_QUORUM_EPOCH =>
+        new BeginQuorumEpochResponseData()
+          .setErrorCode(error.code)
+
+      case ApiKeys.END_QUORUM_EPOCH =>
+        new EndQuorumEpochResponseData()
+          .setErrorCode(error.code)
+
+      case ApiKeys.VOTE =>
+        new VoteResponseData()
+          .setErrorCode(error.code)
+
+      case ApiKeys.FETCH_QUORUM_RECORDS =>
+        new FetchQuorumRecordsResponseData()
+          .setErrorCode(error.code)
+
+      case ApiKeys.FIND_QUORUM =>
+        new FindQuorumResponseData()
+          .setErrorCode(error.code)
+
+      case _ =>
+        throw new AssertionError(s"Unexpected api $key")
+    }
+  }
+
+  private def extractError(response: ApiMessage): Errors = {
+    val code = response match {
+      case res: BeginQuorumEpochResponseData => res.errorCode
+      case res: EndQuorumEpochResponseData => res.errorCode
+      case res: FetchQuorumRecordsResponseData => res.errorCode
+      case res: VoteResponseData => res.errorCode
+      case res: FindQuorumResponseData => res.errorCode
+    }
+    Errors.forCode(code)
+  }
+
+}
+
+object KafkaNetworkChannelTest {
+  val RaftApis = Seq(
```

Comment:
nit: would it be useful to make it part of `ApiKeys` so that other places can also use it?

Reply:
It's not a bad idea. On the other hand, I think the need for this will grow more narrow as the integration with the broker becomes tighter. Hopefully some of the conversions go away and `KafkaNetworkChannel` or whatever becomes the common network layer will work with all request APIs.